[Fleet] Use `GET /uninstall_tokens` in UninstallCommandFlyout

gergoabraham opened the pull request:
## Summary

> **Note**
> To test this, the feature flag `agentTamperProtectionEnabled` should be enabled.

The Uninstall Command Flyout now uses the real `GET /uninstall_tokens` endpoint to fetch the uninstall token for the selected policy, and generates the uninstall commands using the token.

Also, it adds policy ID to the Uninstall Command Flyout.

![image](https://github.com/elastic/kibana/assets/39014407/888ad456-5a64-4289-9626-0890436cd1ba)

Linux/Mac:
![image](https://github.com/elastic/kibana/assets/39014407/3cb73629-d9e5-44cd-a337-f9842fcf1a56)

Windows:
![image](https://github.com/elastic/kibana/assets/39014407/aa296f23-dd71-4b52-8428-1ef767c9ff85)

Loading state:
![image](https://github.com/elastic/kibana/assets/39014407/fde52d37-98e7-44b4-bed2-7bc66656d365)

Error state:
![image](https://github.com/elastic/kibana/assets/39014407/fadbbc9f-5eac-456a-9517-14dc63758c7d)

<details>
<summary>Original screenshots without showing policy ID</summary>

![image](https://github.com/elastic/kibana/assets/39014407/7fd31ee0-e875-43f0-9045-7c6fcbca5257)

Linux/Mac:
![image](https://github.com/elastic/kibana/assets/39014407/cf85144f-c42d-42eb-95f1-563c814cd2b1)

Windows:
![image](https://github.com/elastic/kibana/assets/39014407/389eefcf-646b-4ec8-a74f-18ff6710ea6d)


Loading state:
![image](https://github.com/elastic/kibana/assets/39014407/62646d7e-ec33-4ecc-a66e-44ed05102f0d)

Error state:
![image](https://github.com/elastic/kibana/assets/39014407/40ffbcf4-9666-41aa-9136-55fbca309f1e)
</details>

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

elasticmachine commented:
Pinging @elastic/security-defend-workflows (Team:Defend Workflows)

gergoabraham commented:
@roxana-gheorghe @caitlinbetz @dasansol92 @kevinlog

when I played around with the Uninstall Command Flyout, I found that with a lot of policies, I am not sure if I clicked on the correct one. so before a user sends around the uninstall command, it could be useful for them to be able to double check if the correct policy was selected in a first place.

what do you think about adding the policyID (that the easier) or the policy name (a bit more difficult) to the flyout?

here's a mock:
![image](https://github.com/elastic/kibana/assets/39014407/68439e21-da01-4117-8e38-c127c468d37a)

dasansol92 commented on x-pack/plugins/fleet/public/components/uninstall_command_flyout/uninstall_commands_per_platform.tsx:
```diff
@@ -0,0 +1,69 @@
+}
+
+export const UninstallCommandsPerPlatform: React.FunctionComponent<Props> = ({ token }) => {
+  const [platform, setPlatform] = useState<PLATFORMS_FOR_UNINSTALL>('linuxOrMac');
```
Is this the default we have in the enrollment flyout?

gergoabraham replied:
yes, it's either linux or kubernetes - but we don't have the latter

https://github.com/elastic/kibana/blob/main/x-pack/plugins/fleet/public/components/platform_selector.tsx#L70-L82

dasansol92 commented on x-pack/plugins/fleet/public/hooks/use_request/uninstall_commands.ts:
```diff
@@ -0,0 +1,31 @@
+      policyId,
+      page,
+      perPage,
+    } as GetUninstallTokensRequest['query'],
```
Instead of doing this casting, could you define a const above with these props?

gergoabraham replied:
sure, that's cleaner, thanks!
4afe522fc8c2aeff5c3a1c5d0f416645d8b76810

dasansol92 commented on x-pack/plugins/fleet/public/hooks/use_request/uninstall_commands.ts:
```diff
@@ -0,0 +1,31 @@
+ * 2.0.
+ */
+
+import { uninstallTokensRouteService } from '../../../common/services';
```
Should this file be renamed to: `get_uninstall_tokens` ?

gergoabraham replied:
😅 b16d9f8b8eb

although, without `get_` prefix it fits better with the other files in that folder

dasansol92 commented on x-pack/plugins/fleet/public/components/uninstall_command_flyout/uninstall_command_flyout.test.tsx:
```diff
@@ -58,43 +72,89 @@ describe('UninstallCommandFlyout', () => {
+    });
+  });
+
+  describe('fetching', () => {
```
could be this a bit more specific? For example: `When loading and when error` or something similar. Or maybe you can move the loading `it` to the `describe` above and use this only for error cases.

gergoabraham replied:
sure, thanks for the suggestion!
a241b63a720

dasansol92 reviewed (commented):
This looks awesome! Left a few suggestions, let me know what do you think

dasansol92 reviewed (approved):
LGTM!

nchaulet commented:
@gergoabraham I tried to run that PR and I do not get the button in the agent policy page? did you forget to commit a file?

gergoabraham commented:
@nchaulet, I'm sorry, I forgot to add to the description how to test it.

You need to enable the `agentTamperProtectionEnabled` feature flag for fleet.

nchaulet reviewed (approved):
Tested locally and it works as expected 🚀

gergoabraham commented:
@elasticmachine merge upstream

gergoabraham commented:
> @gergoabraham Good call, I would say display policy Id and we can change it later if needed. In most of the cases, policy Id would be enough I guess and we avoid doing an extra API call at this point.

@dasansol92
added it and updated the screenshots, too.

gergoabraham commented on x-pack/plugins/fleet/public/components/uninstall_command_flyout/uninstall_command_flyout.tsx:
```diff
@@ -110,7 +144,21 @@ export const UninstallCommandFlyout: React.FunctionComponent<UninstallCommandFly
         <EuiSpacer size="l" />
 
-        <CommandsForPlatforms commands={commands} />
+        {policyId && (
```
as discussed offline, now `UninstallCommandFlyout` is accessible only if `policyId` is present: ad0f3cd

it turned out there is a case when it's not present: when creating a new policy, there was the 'Get uninstall tokens' button in the advanced options. not that's hidden on policy creation.

gergoabraham commented:
@elasticmachine merge upstream

gergoabraham commented:
@elasticmachine merge upstream

gergoabraham commented:
@elasticmachine merge upstream

gergoabraham commented:
@elasticmachine merge upstream

gergoabraham commented:
@elasticmachine merge upstream
